### DISCUSSION TITLE
Are JpegEncoder and other format encoders considered to be reusable for multiple images?

### DISCUSSION THREAD

**Discussion by br3aker** [SixLabors/ImageSharp#1711] [Q&A] [answered]
### If yes

There's a bug at least in jpeg:

`JpegEncoder` overwrites grayscale/colored images encoding so encoding combinations of different pixel types wouldn't work properly:

https://github.com/SixLabors/ImageSharp/blob/9baba86a3a08c630e07f349b45b21f94fe2c4805/src/ImageSharp/Formats/Jpeg/JpegEncoder.cs#L66-L83

Steps to reproduce:

```csharp
using var img_grayscale = Image.Load();
using var img_colored = Image.Load();

JpegEncoder encoder = new JpegEncoder { };
img_grayscale.SaveAsJpeg("path", encoder); // encoder is now grascale
img_colored.SaveAsJpeg("path", encoder); // saved as grayscale
```

### If no

Current tests for jpeg assumes encoder can be used multiple times:

https://github.com/SixLabors/ImageSharp/blob/9baba86a3a08c630e07f349b45b21f94fe2c4805/tests/ImageSharp.Tests/Formats/Jpg/JpegEncoderTests.cs#L178-L222

While this currently does not break anything it gives false (or not) promise of being reusable through tests while not being so. Not sure if it wasn't expected so opened a discussion instead of an issue.

**Question/Proposal**: state explicitly in the docs if they are reusable or not and fix reusable usage for all encoders in tests.

**Comment by antonfirsov** [comment, marked as answer]
@br3aker good catch! I think this is a bug, encoder/decoder state should not be altered by encoding. We can implement an equivalent logic in `JpegEncoderCore` instead.

**Reply by br3aker** [reply]
We should check all of them I guess. I might be wrong but I think I saw something state-altering in the png encoder also. Will recheck tommorow and open an issue,

**Reply by JimBobSquarePants** [reply]
Yep, that code and similar should be in the underlying Core encoder which we new up every time.

**Reply by br3aker** [reply]
Looks like every other encoder doesn't suffer from similar issue, opened an issue.
